core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -997,4 +999,27 @@ private[spark] object Executor {
 
   // Used to store executorSource, for local mode only
   var executorSourceLocalModeOnly: ExecutorSource = null
+
+  /**
+   * Whether a `Throwable` thrown from a task is a fatal error. We use this to decide whether to
+   * kill the executor.
+   *
+   * @param shouldDetectNestedFatalError whether to go through the exception chain to check whether
+   *                                     exists a fatal error.
+   * @param depth the current depth of the recursive call. Return `false` when it's greater than 5.
+   *              This is to avoid `StackOverflowError` when hitting a cycle in the exception chain.
+   */
+  def isFatalError(t: Throwable, shouldDetectNestedFatalError: Boolean, depth: Int = 0): Boolean = {
+    if (depth <= 5) {
+      t match {
+        case _: SparkOutOfMemoryError => false
```

Comment:
Just in case, we are sure that OOM cannot be caused by a fatal error, and it cannot present somewhere in the chain?

Reply:
This is an existing behavior. #20014 added SparkOutOfMemoryError to avoid killing the executor when it's not thrown by JVM.